### ISSUE TITLE
etcd3 _load_cluster() does not fail if etcd is down

### ISSUE THREAD

**Issue by mbanck** [zalando/patroni#2180]
**Describe the bug**

I noticed the `dcs_last_seen` flag is being updated by Patroni internally (I added some debugging to print it out) even if etcd is unavailable. The same seems to be not true if I use etcd (i.e. etcdv2)

**To Reproduce**

I added debug logging at the beginning and end of `_load_cluster()` as well, with etcd3, I get:

```
Jan 21 20:46:52 pg1 patroni@11-main2[47010]: 2022-01-21 20:46:52,283 INFO: _load_cluster start
Jan 21 20:46:52 pg1 patroni@11-main2[47010]: 2022-01-21 20:46:52,283 INFO: _load_cluster end
Jan 21 20:46:52 pg1 patroni@11-main2[47010]: 2022-01-21 20:46:52,284 INFO: Updating _last_seen to 1642798012
Jan 21 20:46:52 pg1 patroni@11-main2[47010]: 2022-01-21 20:46:52,289 INFO: no action. I am (pg1), a secondary, and following a leader (pg3)
[at this point I sent SIGSTOP to etcd]
Jan 21 20:47:02 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:02,283 INFO: _load_cluster start
Jan 21 20:47:02 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:02,283 INFO: _load_cluster end
[_load_cluster returns immediately, and _last_seen is updated:]
Jan 21 20:47:02 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:02,284 INFO: Updating _last_seen to 1642798022
Jan 21 20:47:02 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:02,287 INFO: no action. I am (pg1), a secondary, and following a leader (pg3)
Jan 21 20:47:12 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:12,283 INFO: _load_cluster start
Jan 21 20:47:12 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:12,284 INFO: _load_cluster end
Jan 21 20:47:12 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:12,284 INFO: Updating _last_seen to 1642798032
Jan 21 20:47:15 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:12,284 INFO: Lock owner: pg3; I am pg1
Jan 21 20:47:15 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:15,634 ERROR: Request to server http://10.0.3.164:2379 failed: ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=3.333079530353037)")
Jan 21 20:47:15 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:15,634 INFO: Reconnection allowed, looking for another server.
Jan 21 20:47:18 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:18,353 ERROR: Request to server http://10.0.3.164:2379 failed: ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=1.9734664472052827)")
Jan 21 20:47:18 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:18,353 INFO: Reconnection allowed, looking for another server.
Jan 21 20:47:19 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:19,810 ERROR: Request to server http://10.0.3.164:2379 failed: ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=1.2437164426470797)")
Jan 21 20:47:19 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:19,810 INFO: Reconnection allowed, looking for another server.
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,301 ERROR: Request to server http://10.0.3.164:2379 failed: ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=1.7452067360281944)")
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,301 INFO: Reconnection allowed, looking for another server.
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,302 ERROR:
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: Traceback (most recent call last):
Jan 21 20:47:22 pg1 patroni@11-main2[47010]:   File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 564, in wrapper
[...]
Jan 21 20:47:22 pg1 patroni@11-main2[47010]:   File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 233, in _do_http_request
Jan 21 20:47:22 pg1 patroni@11-main2[47010]:     raise etcd.EtcdConnectionFailed('No more machines in the cluster')
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: etcd.EtcdConnectionFailed: No more machines in the cluster
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,305 INFO: no action. I am (pg1), a secondary, and following a leader (pg3)
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,306 WARNING: Loop time exceeded, rescheduling immediately.
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,306 INFO: _load_cluster start
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,307 INFO: _load_cluster end
[it updates _last_seen here again:]
Jan 21 20:47:22 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,307 INFO: Updating _last_seen to 1642798042
Jan 21 20:47:25 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:22,307 INFO: Lock owner: pg3; I am pg1
Jan 21 20:47:25 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:25,647 ERROR: Request to server http://10.0.3.164:2379 failed: ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=3.332729346312893)")
Jan 21 20:47:25 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:25,647 INFO: Reconnection allowed, looking for another server.
Jan 21 20:47:28 pg1 patroni@11-main2[47010]: 2022-01-21 20:47:28,399 ERROR: Request to server http://10.0.3.164:2379 failed: ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=1.9565006268288319)")
```

**Expected behavior**

With etcd2, I get:

```
Jan 21 20:51:19 pg1 patroni@11-main[46974]: 2022-01-21 20:51:19,770 INFO: _load_cluster start
Jan 21 20:51:19 pg1 patroni@11-main[46974]: 2022-01-21 20:51:19,773 INFO: _load_cluster end
Jan 21 20:51:19 pg1 patroni@11-main[46974]: 2022-01-21 20:51:19,773 INFO: Updating _last_seen to 1642798279
Jan 21 20:51:19 pg1 patroni@11-main[46974]: 2022-01-21 20:51:19,778 INFO: no action. I am (pg1), a secondary, and following a leader (pg2)
[at this point I sent SIGSTOP to etcd]
Jan 21 20:51:30 pg1 patroni@11-main[46974]: 2022-01-21 20:51:30,284 INFO: _load_cluster start
[this hangs, contrary to etcd3]
Jan 21 20:51:33 pg1 patroni@11-main[46974]: 2022-01-21 20:51:33,621 WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=0, status=None)) after connection broken by 'ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=3.332642253333082)")': /v2/keys/postgresql-common/11-main/?recursive=true
Jan 21 20:51:36 pg1 patroni@11-main[46974]: 2022-01-21 20:51:36,957 WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=0, status=None)) after connection broken by 'ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=3.333006057267388)")': /v2/keys/postgresql-common/11-main/?recursive=true
Jan 21 20:51:40 pg1 patroni@11-main[46974]: 2022-01-21 20:51:40,295 ERROR: Request to server http://10.0.3.164:2379 failed: MaxRetryError('HTTPConnectionPool(host=\'10.0.3.164\', port=2379): Max retries exceeded with url: /v2/keys/postgresql-common/11-main/?recursive=true (Caused by ReadTimeoutError("HTTPConnectionPool(host=\'10.0.3.164\', port=2379): Read timed out. (read timeout=3.332753581305345)"))')
Jan 21 20:51:40 pg1 patroni@11-main[46974]: 2022-01-21 20:51:40,296 INFO: Reconnection allowed, looking for another server.
Jan 21 20:51:40 pg1 patroni@11-main[46974]: 2022-01-21 20:51:40,296 ERROR: get_cluster
Jan 21 20:51:40 pg1 patroni@11-main[46974]: Traceback (most recent call last):
Jan 21 20:51:40 pg1 patroni@11-main[46974]:   File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 593, in _load_cluster
[...]
Jan 21 20:51:40 pg1 patroni@11-main[46974]:   File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 233, in _do_http_request
Jan 21 20:51:40 pg1 patroni@11-main[46974]:     raise etcd.EtcdConnectionFailed('No more machines in the cluster')
Jan 21 20:51:40 pg1 patroni@11-main[46974]: etcd.EtcdConnectionFailed: No more machines in the cluster
Jan 21 20:51:40 pg1 patroni@11-main[46974]: 2022-01-21 20:51:40,297 ERROR: Error communicating with DCS
Jan 21 20:51:40 pg1 patroni@11-main[46974]: 2022-01-21 20:51:40,299 INFO: DCS is not accessible
Jan 21 20:51:40 pg1 patroni@11-main[46974]: 2022-01-21 20:51:40,300 WARNING: Loop time exceeded, rescheduling immediately.
Jan 21 20:51:40 pg1 patroni@11-main[46974]: 2022-01-21 20:51:40,300 INFO: _load_cluster start
Jan 21 20:51:43 pg1 patroni@11-main[46974]: 2022-01-21 20:51:43,637 WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=0, status=None)) after connection broken by 'ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=3.332795136336548)")': /v2/keys/postgresql-common/11-main/?recursive=true
Jan 21 20:51:46 pg1 patroni@11-main[46974]: 2022-01-21 20:51:46,973 WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=0, status=None)) after connection broken by 'ReadTimeoutError("HTTPConnectionPool(host='10.0.3.164', port=2379): Read timed out. (read timeout=3.3327052503203354)")': /v2/keys/postgresql-common/11-main/?recursive=true
Jan 21 20:51:50 pg1 patroni@11-main[46974]: 2022-01-21 20:51:50,310 ERROR: Request to server http://10.0.3.164:2379 failed: MaxRetryError('HTTPConnectionPool(host=\'10.0.3.164\', port=2379): Max retries exceeded with url: /v2/keys/postgresql-common/11-main/?recursive=true (Caused by ReadTimeoutError("HTTPConnectionPool(host=\'10.0.3.164\', port=2379): Read timed out. (read timeout=3.332756304376138)"))')
Jan 21 20:51:50 pg1 patroni@11-main[46974]: 2022-01-21 20:51:50,310 INFO: Reconnection allowed, looking for another server.
Jan 21 20:51:50 pg1 patroni@11-main[46974]: 2022-01-21 20:51:50,310 ERROR: No more machines in the cluster
Jan 21 20:51:50 pg1 patroni@11-main[46974]: 2022-01-21 20:51:50,310 ERROR: Error communicating with DCS
Jan 21 20:51:50 pg1 patroni@11-main[46974]: 2022-01-21 20:51:50,312 INFO: DCS is not accessible
Jan 21 20:51:50 pg1 patroni@11-main[46974]: 2022-01-21 20:51:50,313 WARNING: Loop time exceeded, rescheduling immediately.
```

And `_last_seen` is not updated anymore, which I think is how it should be.

**Environment**
- Patroni version: 2.1.2
- PostgreSQL version: 11.14
- DCS (and its version): etcd-3.3.25

**Comment by CyberDem0n** [comment]
The difference between etcd and etcd3 implementations is that the last one is watching key changes (long-polling connection) under a certain prefix, while the former is polling keys in every HA loop.

Polling mode produces more load on etcd nodes, but the good side-effect that it allows immediately detecting cases when etcd cluster is not accessible.

Watching is only possible with etcd3 and it allows to build a local cache of keys that belong to a given cluster. 
I.e, the _load_cluster() isn't doing any HTTP calls, but simply copying data from the local cache if it is "ready".
Ready - it means that we fetched the initial keys and started a watcher thread that receiving all changes from a given revision (something similar to transaction id).
In case of cache not being ready _load_cluster() waits up to retry_timeout and gives up.

The long-polling connection socket is configured with TCP keepalives and timeouts that are aligned with ttl, loop_wait, and retry_timeout and is supposed to detect the case when etcd node "silently" disappeared.
I don't know how you performed the test,  but for example if you simply "pause" etcd with SIGSTOP, the watcher connection won't be able to "detect" it and you get described behavior. 

Switching back to polling with etcd3 isn't a good option. The best thing we can do - is to interrupt the watcher if some other interaction (for example, every node calls lease_keepalive() for every heart-beat) with etcd cluster failed for all nodes. It will make the `_last_seen` to stop updating, but only from the next HA loop, which IMO is a good compromise.

**Comment by mbanck** [comment]
Thanks for the explanation.

> I don't know how you performed the test, but for example if you simply "pause" etcd with SIGSTOP, the watcher connection 
> won't be able to "detect" it and you get described behavior.

Right, I did that. I agree this is somewhat synthetic, but I'm debugging a etcd3-based test setup where I saw the same behavior; however the etcd3 is proxy'd via an nginx ingress here (the etcd3 nodes are in a k8s cluster, while patroni is running on regular nodes outside, and nginx provides a service-endpoint, that wasn't my idea and I dont' operate it), which apparently means the connection isn't lost, but one gets error messages. I guess due to that `_last_seen` is being updated.

Here's a snippet from that setup:

```
2022-01-21 15:10:23,411 INFO: Updating _last_seen to 1642774223
2022-01-21 15:10:23,441 ERROR: refresh_lease
Traceback (most recent call last):
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd3.py", line 617, in refresh_lease
    return self.retry(self._do_refresh_lease)
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd.py", line 445, in retry
    return retry(*args, **kwargs)
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/utils.py", line 334, in __call__
    return func(*args, **kwargs)
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd3.py", line 605, in _do_refresh_lease
    if self._lease and not self._client.lease_keepalive(self._lease, retry):
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd3.py", line 322, in lease_keepalive
    return self.call_rpc('/lease/keepalive', {'ID': ID}, retry).get('result', {}).get('TTL')
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd3.py", line 558, in call_rpc
    ret = super(PatroniEtcd3Client, self).call_rpc(method, fields, retry)
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd3.py", line 262, in call_rpc
    return self.api_execute(self.version_prefix + method, self._MPOST, fields)
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd.py", line 257, in api_execute
    return self._handle_server_response(response)
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd3.py", line 220, in _handle_server_response
    _raise_for_data(data, response.status)
  File "/srv/container/postgres/.local/lib/python3.6/site-packages/patroni/dcs/etcd3.py", line 136, in _raise_for_data
    raise err(code, error, status_code)
patroni.dcs.etcd3.Unknown: <Unknown error: '<html>^M
<head><title>404 Not Found</title></head>^M
<body>^M
<center><h1>404 Not Found</h1></center>^M
<hr><center>nginx</center>^M
</body>^M
</html>', code: 2>
2022-01-21 15:10:23,441 INFO: following a different leader because i am not the healthiest node
2022-01-21 15:10:33,273 INFO: Updating _last_seen to 1642774233
```

I will try to debug this further.

> Switching back to polling with etcd3 isn't a good option. The best thing we can do - is to interrupt the watcher if some other 
> interaction (for example, every node calls lease_keepalive() for every heart-beat) with etcd cluster failed for all nodes. It will 
> make the _last_seen to stop updating, but only from the next HA loop, which IMO is a good compromise.

Sounds good.

**Comment by CyberDem0n** [comment]
> etcd3 is proxy'd via an nginx ingress

I.e., there is a single endpoint for all etcd nodes? That would be a very non-optimal setup, very similar to Consul which we can work only via the local agent. When working directly with etcd members Patroni can better calculate timeouts/retries and prefers switching to another node if retrying requests.

> patroni.dcs.etcd3.Unknown: <Unknown error: '<html>^M

This is weird. The error comes from nginx, not etcd. Could it be that nginx is configured to proxy only certain locations and/or request methods?

**Comment by ADMDevMit** [comment]
I also ran into this problem.

How to reproduce:
```
node1 etcd+patroni, leader
node2 etcd+patroni, secondary
node3 etcd (witness), etcd cluster leader
```
node2:
`systemctl stop etcd`
<details>
  <summary><i>patroni connects to another etcd node (node1)</i></summary>
<pre>
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
ERROR: watchprefix failed: <Unavailable error: 'transport is closing', code: 14>
ERROR: Request to server https://node2.example.com:2379 failed: MaxRetryError("HTTPSConnectionPool(host='node2.example.com', port=2379): Max retries exceeded with url: /v3/kv/range (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fed13d3cc70>: Failed to establish a new connection: [Errno 111] Connection refused'))")
INFO: Reconnection allowed, looking for another server.
INFO: Retrying on https://node1.example.com:2379
INFO: Selected new etcd server https://node1.example.com:2379
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
</pre>
</details>

`systemctl start etcd`
_No changes._

It turns out that both patroni node1 and node2 are connected to node1 etcd.
Since etcd is already running on node2, etcd cluster is healthy.

node1:
`systemctl stop etcd`
<details>
<summary>patroni node1 log</summary>
<pre>
INFO: no action. I am (node1), the leader with the lock
INFO: no action. I am (node1), the leader with the lock
ERROR: watchprefix failed: <Unavailable error: 'transport is closing', code: 14>
ERROR: Request to server https://node1.example.com:2379 failed: MaxRetryError("HTTPSConnectionPool(host='node1.example.com', port=2379):
 Max retries exceeded with url: /v3/kv/range (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f423cd7bc70>: Failed to establish a
 new connection: [Errno 111] Connection refused'))")
INFO: Reconnection allowed, looking for another server.
INFO: Retrying on https://node3.example.com:2379
INFO: Selected new etcd server https://node3.example.com:2379
ERROR: Failed to get list of machines from https://node3.example.com:2379/v3: <Unknown error: '404 page not found', code: 2>
ERROR: Failed to get list of machines from https://node1.example.com:2379/v3: MaxRetryError("HTTPSConnectionPool(host='node1.example.com
', port=2379): Max retries exceeded with url: /v3/cluster/member/list (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f423c4d670
0>: Failed to establish a new connection: [Errno 111] Connection refused'))")
ERROR: KVCache.run <Unknown error: '404 page not found', code: 2>
ERROR: KVCache.run <Unknown error: '404 page not found', code: 2>
ERROR: KVCache.run <Unknown error: '404 page not found', code: 2>
ERROR: get_cluster
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd3.py", line 644, in _load_cluster
    for node in self._client.get_cluster():
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd3.py", line 552, in get_cluster
    self._wait_cache(self._etcd3._retry.deadline)
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd3.py", line 546, in _wait_cache
    raise RetryFailedError('Exceeded retry deadline')
patroni.utils.RetryFailedError: 'Exceeded retry deadline'
ERROR: Error communicating with DCS
INFO: Demoting self (offline)
ERROR: KVCache.run <Unknown error: '404 page not found', code: 2>
INFO: closed patroni connection to the postgresql cluster
INFO: postmaster pid=94
INFO: demoted self because DCS is not accessible and i was a leader
WARNING: Loop time exceeded, rescheduling immediately.
ERROR: KVCache.run <Unknown error: '404 page not found', code: 2>
ERROR: KVCache.run <Unknown error: '404 page not found', code: 2>
ERROR: 'Exceeded retry deadline'
ERROR: Error communicating with DCS
INFO: establishing a new patroni connection to the postgres cluster
INFO: DCS is not accessible
</pre>
</details>

<details>
<summary>patroni node2 log</summary>
<pre>
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
ERROR: watchprefix failed: <Unavailable error: 'transport is closing', code: 14>
ERROR: Request to server https://node1.example.com:2379 failed: MaxRetryError("HTTPSConnectionPool(host='node1.example.com', port=2379):
 Max retries exceeded with url: /v3/kv/range (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa03053f670>: Failed to establish a
 new connection: [Errno 111] Connection refused'))")
INFO: Reconnection allowed, looking for another server.
INFO: Retrying on https://node2.example.com:2379
INFO: Selected new etcd server https://node2.example.com:2379
ERROR: Invalid auth token: nzVsCwqCNwPvoFDe.1337
INFO: Trying to authenticate on Etcd...
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: no action. I am (node2), a secondary, and following a leader (node1)
INFO: Got response from node1 https://node1.example.com:8008/patroni: {"state": "running", "postmaster_start_time": "2022-01-25 23:02:44.
414466+03:00", "role": "replica", "server_version": 130005, "xlog": {"received_location": 167772720, "replayed_location": 167772720, "replayed_timestamp": null,
 "paused": false}, "timeline": 33, "replication": [{"usename": "replicator_user", "application_name": "node2", "client_addr": "[node2_ip]", "state": "streaming", "
sync_state": "async", "sync_priority": 0}], "cluster_unlocked": true, "dcs_last_seen": 1643140960, "database_system_identifier": "7056382928450744357", "patroni
": {"version": "2.1.2", "scope": "scope"}}
WARNING: Could not activate Linux watchdog device: "Can't open watchdog device: [Errno 2] No such file or directory: '/dev/watchdog'"
INFO: promoted self to leader by acquiring session lock
INFO: cleared rewind state after becoming the leader
INFO: Lock owner: node2; I am node2
INFO: updated leader lock during promote
INFO: no action. I am (node2), the leader with the lock
INFO: no action. I am (node2), the leader with the lock
</pre>
</details>

**Comment by ADMDevMit** [comment]
@CyberDem0n, unfortunately, this patch did not solve the problem.

But I found the source of this error and how to fix it.
First of all, I want to apologize, in the previous message I forgot to indicate that node3 (with etcd-witness) is running on Windows.
That was the main problem.

When starting etcd with the `--config-file` parameter, the use of environment variables and other command-line parameters is excluded, etcd ignores them.
Thus, it turned out that the etcd server processed requests to the HTTP API in API v2 mode (I looked in the logs in debug mode), and did not find the paths/keys of API v3.
(Should I create a separate issue about this?)
The problem disappeared when I ran etcd with parameters on the command line, instead of `--config-file`. The server began to support API v3.

Now everything is working fine and in the version without a patch, patroni switches to another working etcd server when the current one fails.

But this does not exclude the fact that there is a bug, as a result of which, when patroni receiving a response '404 not found' from the etcd server, patroni does not continue to search and connect for other available etcd servers.
